core-client/src/transports/duplex.rs

```diff
@@ -0,0 +1,126 @@
+//! Duplex transport
+
+use failure::format_err;
+use futures::prelude::*;
+use futures::sync::{mpsc, oneshot};
+use jsonrpc_core::Id;
+use log::debug;
+use serde_json::Value;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+
+use super::RequestBuilder;
+use crate::{RpcChannel, RpcError, RpcMessage};
+
+/// The Duplex handles sending and receiving asynchronous
+/// messages through an underlying transport.
+pub struct Duplex<TSink, TStream> {
+	request_builder: RequestBuilder,
+	queue: HashMap<Id, oneshot::Sender<Result<Value, RpcError>>>,
+	sink: TSink,
+	stream: TStream,
+	channel: Option<mpsc::Receiver<RpcMessage>>,
+	outgoing: VecDeque<String>,
+}
+
+impl<TSink, TStream> Duplex<TSink, TStream> {
+	/// Creates a new `Duplex`.
+	fn new(sink: TSink, stream: TStream, channel: mpsc::Receiver<RpcMessage>) -> Self {
+		Duplex {
+			request_builder: RequestBuilder::new(),
+			queue: HashMap::new(),
+			sink,
+			stream,
+			channel: Some(channel),
+			outgoing: VecDeque::new(),
+		}
+	}
+
+	/// Creates a new `Duplex`, along with a channel to communicate
+	pub fn with_channel(sink: TSink, stream: TStream) -> (Self, RpcChannel) {
```

Comment:
The method is a bit confusing, but seeing `with_channel` I think I was expecting the channel to be passed as parameter instead, I'd swap the names.

Reply:
I've pulled this out to a public function `duplex`. Not sure about the name but perhaps more idiomatic for creating the `(Duplex, RpcChannel)` tuple.